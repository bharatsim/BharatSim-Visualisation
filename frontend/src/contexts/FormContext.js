import React, { useContext } from 'react';

const FormContext = React.createContext({});

const FormProvider = FormContext.Provider;

const useFormContext = () => useContext(FormContext);

export { FormContext, FormProvider, useFormContext };
