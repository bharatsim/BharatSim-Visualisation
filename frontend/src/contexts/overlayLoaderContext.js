import React from 'react';

const overlayLoaderContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderContextProvider = overlayLoaderContext.Provider;

export { overlayLoaderContext, OverlayLoaderContextProvider };
