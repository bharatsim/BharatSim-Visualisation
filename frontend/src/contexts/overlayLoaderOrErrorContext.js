import React from 'react';

const overlayLoaderOrErrorContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderOrErrorContextProvider = overlayLoaderOrErrorContext.Provider;

export { overlayLoaderOrErrorContext, OverlayLoaderOrErrorContextProvider };
