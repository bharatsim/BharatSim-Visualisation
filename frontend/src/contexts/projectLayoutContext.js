import React from 'react';

const projectLayoutContext = React.createContext({
  projectMetadata: {
    id: undefined,
    name: undefined,
  },
  selectedDashboardMetadata: {
    name: undefined,
    _id: undefined,
  },
  addDashboard: null,
  deleteDashboard: null,
});

const ProjectLayoutProvider = projectLayoutContext.Provider;

export { projectLayoutContext, ProjectLayoutProvider };
