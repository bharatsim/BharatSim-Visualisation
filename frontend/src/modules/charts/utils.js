import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
