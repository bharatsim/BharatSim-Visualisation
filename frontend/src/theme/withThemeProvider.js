import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
