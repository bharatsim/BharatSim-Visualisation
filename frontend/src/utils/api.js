import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
