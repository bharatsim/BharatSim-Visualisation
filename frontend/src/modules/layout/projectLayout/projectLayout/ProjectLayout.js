import React, { useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';

import { Box } from '@material-ui/core';
import DashboardNavbar from '../sideDashboardNavbar/DashboardNavbar';

import useProjectLayoutStyle from './projectLayoutCSS';

import { api } from '../../../../utils/api';
import { ProjectLayoutProvider } from '../../../../contexts/projectLayoutContext';
import { ChildrenPropTypes } from '../../../../commanPropTypes';

function ProjectLayout({ children }) {
  const classes = useProjectLayoutStyle();
  const { id: projectId } = useParams();
  const history = useHistory();

  const [projectMetadata, setProjectMetadata] = useState({
    id: null,
    name: 'untitled project',
  });

  const [dashboards, setDashboards] = useState(undefined);
  const [selectedDashboard, setSelectedDashboard] = useState(0);

  useEffect(() => {
    async function fetchData() {
      if (projectId) {
        const fetchedProject = await api.getProject(projectId);
        const { _id, name } = fetchedProject.project;
        setProjectMetadata({ id: _id, name });
      }
    }
    fetchData();
  }, [projectId]);

  useEffect(() => {
    async function fetchData() {
      if (projectId) {
        const fetchedDashboards = await api.getAllDashBoardByProjectId(projectId);
        setDashboards(fetchedDashboards.dashboards);
      }
    }
    fetchData();
  }, [projectId]);

  useEffect(() => {
    if (projectId) {
      history.push(`/projects/${projectId}/dashboard`);
    }
  }, [projectId, selectedDashboard]);

  useEffect(() => {
    if (dashboards && dashboards.length === 0) {
      history.replace({ pathname: `/projects/${projectId}/create-dashboard` });
    }
  }, [dashboards, projectId]);

  function addDashboard(dashboard) {
    const prevDashboards = dashboards || [];
    setDashboards([...prevDashboards, dashboard]);
    const nextDashboardIndex = prevDashboards.length;
    setSelectedDashboard(nextDashboardIndex);
  }

  function deleteDashboard(selectedDashboardId) {
    const updatedDashboards = dashboards.filter(
      ({ _id: dashboardId }) => dashboardId !== selectedDashboardId,
    );
    setDashboards(updatedDashboards);
  }

  if (projectId && (!dashboards || !projectMetadata.id)) {
    return null;
  }

  return (
    <Box className={classes.layoutContainer}>
      <ProjectLayoutProvider
        value={{
          projectMetadata,
          selectedDashboardMetadata: (dashboards && dashboards[selectedDashboard]) || {},
          addDashboard,
          deleteDashboard,
        }}
      >
        <Box className={classes.sideBarLayout}>
          <DashboardNavbar
            navItems={dashboards ? dashboards.map((dashboard) => dashboard) : []}
            value={selectedDashboard}
            setNavTab={setSelectedDashboard}
          />
        </Box>
        <Box className={classes.viewContainer}>{children}</Box>
      </ProjectLayoutProvider>
    </Box>
  );
}

ProjectLayout.propTypes = {
  children: ChildrenPropTypes.isRequired,
};

export default ProjectLayout;
