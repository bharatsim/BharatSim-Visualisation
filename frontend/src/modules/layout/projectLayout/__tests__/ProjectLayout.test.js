import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import * as router from 'react-router-dom';
import { fireEvent } from '@testing-library/dom';
import { api } from '../../../../utils/api';
import withThemeProvider from '../../../../theme/withThemeProvider';
import ProjectLayout from '../projectLayout/ProjectLayout';
import { projectLayoutContext } from '../../../../contexts/projectLayoutContext';

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();

function DummyComponent() {
  const { projectMetadata, selectedDashboardMetadata, addDashboard } = useContext(
    projectLayoutContext,
  );
  return (
    <div>
      <div>ProjectLayout Child</div>
      <button onClick={() => addDashboard({ _id: 'id', name: 'dashboard-name' })} type="button">
        add dashboard
      </button>
      {JSON.stringify({ projectMetadata, selectedDashboardMetadata }, null, 2)}
    </div>
  );
}

jest.mock('../../../../utils/api', () => ({
  api: {
    getProject: jest.fn().mockResolvedValue({ project: { name: 'name', _id: 'id' } }),
    getAllDashBoardByProjectId: jest
      .fn()
      .mockResolvedValue({ dashboards: [{ name: 'd_name', _id: 'd_id' }] }),
  },
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
  useHistory: () => ({
    push: mockHistoryPush,
    replace: mockHistoryReplace,
  }),
}));

describe('Project', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const Component = withThemeProvider(() => (
    <ProjectLayout>
      <DummyComponent />
    </ProjectLayout>
  ));

  it('should render child without any api call if project id is undefined', async () => {
    router.useParams.mockReturnValue({ id: undefined });

    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });

  it('should render child with api call if project id is present', async () => {
    router.useParams.mockReturnValue({ id: 1 });
    api.getProject.mockResolvedValue({ project: { _id: 1, name: 'project1' } });

    const { findByText } = render(<Component />);

    await findByText('ProjectLayout Child');

    expect(api.getAllDashBoardByProjectId).toHaveBeenCalledWith(1);
    expect(api.getProject).toHaveBeenCalledWith(1);
  });

  it('should create tab for dashboard in sidebar panel', async () => {
    router.useParams.mockReturnValue({ id: 1 });
    api.getProject.mockResolvedValue({ project: { _id: 1, name: 'project1' } });

    const { findByText } = render(<Component />);
    await findByText('ProjectLayout Child');

    expect(await findByText('d_name')).not.toBeNull();
  });

  it('should navigate to create dashboard page if dashboards are empty', async () => {
    router.useParams.mockReturnValue({ id: 1 });
    api.getAllDashBoardByProjectId.mockResolvedValue({ dashboards: [] });

    const { findByText } = render(<Component />);

    await findByText('ProjectLayout Child');

    expect(mockHistoryReplace).toHaveBeenCalledWith({ pathname: '/projects/1/create-dashboard' });
  });

  it('should create tab for dashboard in sidebar panel on click of add dashboard', async () => {
    router.useParams.mockReturnValue({ id: 1 });

    const { getByText, findByText } = render(<Component />);

    await findByText('ProjectLayout Child');

    fireEvent.click(getByText('add dashboard'));

    expect(getByText('dashboard-name')).not.toBeNull();
  });
});
