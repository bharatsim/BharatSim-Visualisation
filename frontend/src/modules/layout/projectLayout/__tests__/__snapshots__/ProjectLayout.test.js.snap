// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project should render child without any api call if project id is undefined 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-5 makeStyles-layoutContainer-2"
  >
    <div
      class="MuiBox-root MuiBox-root-6 makeStyles-sideBarLayout-3"
    >
      <div
        class="MuiBox-root MuiBox-root-13"
      >
        <div
          class="MuiTabs-root MuiTabs-vertical"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer MuiTabs-flexContainerVertical"
              role="tablist"
            />
            <span
              class="PrivateTabIndicator-root-17 PrivateTabIndicator-colorSecondary-19 MuiTabs-indicator makeStyles-indicator-12 PrivateTabIndicator-vertical-20"
              style="top: 0px; height: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-14 makeStyles-viewContainer-4"
    >
      <div>
        <div>
          ProjectLayout Child
        </div>
        <button
          type="button"
        >
          add dashboard
        </button>
        {
  "projectMetadata": {
    "id": null,
    "name": "untitled project"
  },
  "selectedDashboardMetadata": {}
}
      </div>
    </div>
  </div>
</div>
`;
